src/http/client.rs

```diff
@@ -283,9 +283,19 @@ impl EspHttpConnection {
     pub fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspError> {
         self.assert_response();
 
-        Self::check(unsafe {
-            esp_http_client_read_response(self.raw_client, buf.as_mut_ptr() as _, buf.len() as _)
-        })
+        let result = Self::check(unsafe {
```

Comment:
Hmmmm. I was thinking that the loop which retries on `EAGAIN` would be _inside_ the `read` method, rather than pushing it back to the shoulders of the user?

Reply:
Hmm yeah, I just re-read the documentation of `std::io::Read::read` and noticed that it doesn't encourage either way specifically but leaves it completely open to decide how to do it:
> This function does not provide any guarantees about whether it blocks waiting for data, but if an object needs to block for a read and cannot, it will typically signal this via an [Err](https://doc.rust-lang.org/std/result/enum.Result.html#variant.Err) return value.

I suppose it's a question of design philosophy. Either you hide complexity from the user and make it simpler for them to use your API or you pass the complexity on but give them more choice to handle things their way. In this case maybe they don't want to retry immediately but instead do something else first.

I think it is quite clear which one I prefer in this context (embedded Rust) but in the end I would give you the final say.

Just one more thing to consider is what would happen if the server keeps the connection open but just doesn't send any more data? Would this effectively deadlock the thread?
You can specify a timeout on the EspHttpConnection but from the limited testing I did for this issue it seemed like this only determines after which time period an EAGAIN is returned.